controllers/vaultsecret_controller.go

```diff
@@ -0,0 +1,224 @@
+
+func (r *VaultSecretReconciler) formatK8sSecret(instance *redhatcopv1alpha1.VaultSecret, data interface{}) (*corev1.Secret, error) {
+
+	stringData := make(map[string]string)
+	for k, v := range instance.Spec.TemplatizedK8sSecret.StringData {
+
+		tpl, err := template.New("").Funcs(utilstemplates.AdvancedTemplateFuncMap(r.GetRestConfig(), r.Log)).Parse(v)
+		if err != nil {
+			r.Log.Error(err, "unable to create template", "instance", instance)
+			return nil, err
+		}
+
+		var b bytes.Buffer
+		err = tpl.Execute(&b, data)
+		if err != nil {
+			r.Log.Error(err, "unable to execute template", "instance", instance)
+			return nil, err
+		}
+
+		stringData[k] = b.String()
+	}
+
+	k8sSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        instance.Spec.TemplatizedK8sSecret.Name,
+			Namespace:   instance.Namespace,
+			Annotations: instance.Spec.TemplatizedK8sSecret.Annotations,
+			Labels:      instance.Spec.TemplatizedK8sSecret.Labels,
+		},
+		StringData: stringData,
+		Type:       corev1.SecretType(instance.Spec.TemplatizedK8sSecret.Type),
+	}
+
+	ctrl.SetControllerReference(instance, k8sSecret, r.GetScheme())
+
+	return k8sSecret, nil
+}
+
+func (r *VaultSecretReconciler) manageReconcileLogic(ctx context.Context, instance *redhatcopv1alpha1.VaultSecret) error {
+
+	mergedMap := make(map[string]interface{})
+	for _, kvSecret := range instance.Spec.KVSecrets {
+		vaultClient, err := kvSecret.Authentication.GetVaultClient(ctx, instance.Namespace)
+		if err != nil {
+			r.Log.Error(err, "unable to create vault client", "instance", instance)
+			return err
+		}
+
+		ctx = context.WithValue(ctx, "vaultClient", vaultClient)
+		vaultEndpoint := vaultutils.NewVaultEndpointObj(&kvSecret)
+
+		data, ok, _ := vaultEndpoint.Read(ctx)
+		if !ok {
+			return errors.New("unable to read Key/Value Secret from Vault for " + kvSecret.GetPath())
+		}
+
+		mergedMap[kvSecret.Name] = data
+	}
+
+	k8sSecret, err := r.formatK8sSecret(instance, mergedMap)
+	if err != nil {
+		r.Log.Error(err, "unable to format k8s secret", "instance", instance)
+		return err
+	}
+
+	existingK8sSecret := &corev1.Secret{}
+	err = r.GetClient().Get(ctx, types.NamespacedName{
+		Namespace: instance.GetNamespace(),
+		Name:      k8sSecret.GetName(),
+	}, existingK8sSecret)
+
+	if err != nil {
+		// doesn't exist yet so create it
+		err := r.GetClient().Create(ctx, k8sSecret)
+		if err != nil {
+			r.Log.Error(err, "unable to create k8s secret", "secret", k8sSecret)
+			return err
+		}
+	} else {
+		// already exists, update if needed
+		err := r.GetClient().Update(ctx, k8sSecret)
+		if err != nil {
+			r.Log.Error(err, "unable to update k8s secret", "secret", k8sSecret)
+			return err
+		}
+	}
+
+	now := metav1.NewTime(time.Now())
+	instance.Status.LastVaultSecretUpdate = &now
+
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *VaultSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	needsCreation := predicate.Funcs{
```

Comment:
there is n need to do this, the update event will be sent to you if there is a change anywhere, which is what you are checking here.